evaluation-tasks/util/s3.py

```diff
@@ -0,0 +1,153 @@
+
+    def requires(self):
+        return EnsureBucket(self.bucket, self.region)
+
+    @property
+    def name(self):
+        return type(self).__name__
+
+    def run(self):
+        tarfile = f"{self.task_name}.tar.gz"
+        pathtarfile = f"{self.workdir}/{tarfile}"
+        client = S3Client()
+        s3cache = os.path.join(f"s3://{self.bucket}/", tarfile)
+        if client.exists(s3cache):
+            # TODO: Create a new task to untar this a move it out of
+            # workdir into the main evaluation dir so it is good to go.
+            client.get(s3cache, pathtarfile)
+        else:
+            # If you yield FinalizeCorpus, this task is suspended
+            # and FinalizeCorpus is run, and a LocalFileTarget is
+            # returned.
+            finalize_corpus = yield self.next_task()
+
+            # Tar the file
+            devnull = open(os.devnull, "w")
+            ret = subprocess.call(
+                [
+                    "tar",
+                    "zcvf",
+                    pathtarfile,
+                    # Unfortunately, we have to hardcode
+                    # FinalizeCorpus.workdir()
+                    # because we don't have a Task
+                    # just a Target in finalize_corpus.
+                    self.task_name,
+                ],
+                stdout=devnull,
+                stderr=devnull,
+            )
+            # Make sure the return code is 0 and the command was successful.
+            assert ret == 0
+
+            # Cache to S3
+            print("Putting file to S3")
+            client.put_multipart(pathtarfile, s3cache)
+
+        with self.output().open("w") as outfile:
+            pass
+
+
+class EnsureBucket(WorkTask):
+    """
+    Ensure the S3 bucket exists and is readable.
+
+    This S3 code is pretty gnarly, but I'm not sure it can be made
+    any cleaner.
+    """
+
+    bucket = luigi.Parameter()
+    region = luigi.Parameter()
+
+    @property
+    def name(self):
+        return type(self).__name__
+
+    def run(self):
+        check_bucket(self.bucket, self.region)
+        with self.output().open("w") as outfile:
+            pass
+
+
+def can_access_bucket(s3: boto3.session.Session, bucket: str) -> str:
+    """
+    Checks access to bucket
+
+    Args:
+        s3: s3 connection
+        bucket: name of bucket to check
+
+    Returns:
+        boolean indicating whether bucket can be accessed
+
+    """
+    try:
+        s3.head_bucket(Bucket=bucket)
+        return True
+    except ClientError:
+        return False
+
+
+def check_bucket(bucket: str, region: str):
+    """
+    Attempt to access an S3 bucket. If it doesn't exist,
+    then it is created. An error is raised if the operation
+    was unsuccessful.
+
+    Args:
+        bucket: name of bucket
+        region: region the bucket exists in
+    """
+
+    # TODO: test this when region is set to None and update README on how that works.
```

Comment:
region None wasn't working for me

Reply:
Me neither

Reply:
#11 -- check this when implementing S3 script